api/sda/sda.go

```diff
@@ -206,90 +195,106 @@ func Download(c *gin.Context) {
+		log.Errorf("failed to convert start coordinate %d to integer, %s", start, err)
+		c.String(http.StatusBadRequest, "startCoordinate must be an integer")
 
-		return nil, err
+		return
 	}
-	log.Debugf("file stream for %s constructed", file)
-
-	return c4ghr, nil
-}
-
-// parseCoordinates takes query param coordinates and converts them to
-// Crypt4GH reader format
-var parseCoordinates = func(r *http.Request) (*headers.DataEditListHeaderPacket, error) {
+	end, err := strconv.ParseInt(qEnd, 10, 0)
+	if err != nil {
+		log.Errorf("failed to convert end coordinate %d to integer, %s", end, err)
 
-	coordinates := &headers.DataEditListHeaderPacket{}
+		c.String(http.StatusBadRequest, "endCoordinate must be an integer")
 
-	// Get query params
-	qStart := r.URL.Query().Get("startCoordinate")
-	qEnd := r.URL.Query().Get("endCoordinate")
+		return
+	}
+	if end < start {
+		log.Errorf("endCoordinate=%d must be greater than startCoordinate=%d", end, start)
 
-	// Parse and verify coordinates are valid
-	if len(qStart) > 0 && len(qEnd) > 0 {
-		start, err := strconv.ParseUint(qStart, 10, 64)
-		if err != nil {
-			log.Errorf("failed to convert start coordinate %d to integer, %s", start, err)
+		c.String(http.StatusBadRequest, "endCoordinate must be greater than startCoordinate")
 
-			return nil, errors.New("startCoordinate must be an integer")
-		}
-		end, err := strconv.ParseUint(qEnd, 10, 64)
-		if err != nil {
-			log.Errorf("failed to convert end coordinate %d to integer, %s", end, err)
-
-			return nil, errors.New("endCoordinate must be an integer")
-		}
-		if end < start {
-			log.Errorf("endCoordinate=%d must be greater than startCoordinate=%d", end, start)
+		return
+	}
 
-			return nil, errors.New("endCoordinate must be greater than startCoordinate")
-		}
-		// API query params take a coordinate range to read "start...end"
-		// But Crypt4GHReader takes a start byte and number of bytes to read "start...(end-start)"
-		bytesToRead := end - start
-		coordinates.NumberLengths = 2
-		coordinates.Lengths = []uint64{start, bytesToRead}
+	if start == 0 && end == 0 {
+		c.Header("Content-Length", fmt.Sprint(fileDetails.DecryptedSize))
 	} else {
-		coordinates = nil
+		// Calculate how much we should read (if given)
+		togo := end - start
+		c.Header("Content-Length", fmt.Sprint(togo))
 	}
 
-	return coordinates, nil
+	err = sendStream(c4ghr, c.Writer, start, end)
+	if err != nil {
+		log.Errorf("error occurred while sending stream: %v", err)
+		c.String(http.StatusInternalServerError, "an error occurred")
+
+		return
+	}
 }
 
-// sendStream streams file contents from a reader
-var sendStream = func(w http.ResponseWriter, file io.Reader) {
-	log.Debug("begin data stream")
+// sendStream
+// used from: https://github.com/neicnordic/crypt4gh/blob/master/examples/reader/main.go#L48C1-L113C1
+var sendStream = func(reader *streaming.Crypt4GHReader, writer http.ResponseWriter, start, end int64) error {
 
```

Comment:
Could this possibly be simplified by using `reader.Seek` at start and then use of `io.CopyN`? (Will still need to loop I think, but could be a bit nicer).

Reply:
i will try to improve that when switching to `neicnordic/sensitive-data-archive`, till then I will leave as is https://github.com/neicnordic/sensitive-data-archive/issues/410